src/components/TableDialog.vue

```diff
@@ -157,6 +187,88 @@ export default Vue.extend({
         this.fileName = this.fileName || getFileName(file);
         this.fileUploadError = null;
       }
+
+      interface TypeScore {
+        strings: Set<string>;
+        number: number;
+        date: number;
+        total: number;
+      }
+      const columnTypes = new Map<string, TypeScore>();
+
+      Papa.parse(file, {
+        header: true,
+        skipEmptyLines: true,
+        step(row: ParseResult<{}>) {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const data = row.data as { [key: string]: any };
+
+          Object.keys(data).forEach((key: string) => {
+            if (!columnTypes.has(key)) {
+              columnTypes.set(key, {
+                strings: new Set<string>(),
+                number: 0,
+                date: 0,
+                total: 0,
+              });
+            }
+
+            const entry = columnTypes.get(key);
+            if (entry === undefined) {
+              throw new Error('impossible');
+            }
+
+            // Pass the value through a gauntlet of heuristics to see what types
+            // it can convert to.
+            const value = data[key];
+            entry.total += 1;
+
+            // Test for categoriness by counting up all the unique strings in
+            // the column.
+            entry.strings.add(value);
+
+            // See if the value can be converted to a number. Empty strings
+            // convert to 0, so those are excluded specifically.
+            if (value !== '' && !Number.isNaN(Number(value))) {
+              entry.number += 1;
+            }
+
+            // See if the value looks like a date.
+            if (dayjs(value).isValid()) {
+              entry.date += 1;
+            }
+          });
+        },
+
+        complete: () => {
+          const typeRecs = new Map<string, MultinetType>();
+
+          columnTypes.forEach((entry, field) => {
+            const isKey = field === '_key' || field === '_from' || field === '_to';
+            const category = entry.strings.size <= 10;
+            const number = entry.number === entry.total;
+            const date = entry.date === entry.total;
+
+            let rec: MultinetType = 'label';
+            if (isKey) {
+              rec = 'label';
```

Comment:
Why have this initial if statement? Won't this will always be a noop?

Reply:
Also, what is `rec` supposed to mean? I think either a better variable name or a comment explaining this would help

Reply:
I think `rec` is short for recommendation. I think that's reasonable. 

Also, the first statement has to exists, because the 4 boolean expressions above can all evaluate to false, which would cause issues. The default if we can't figure out the type should probably be label

Reply:
> I think `rec` is short for recommendation. I think that's reasonable.

I agree.

> Also, the first statement has to exists, because the 4 boolean expressions above can all evaluate to false, which would cause issues. The default if we can't figure out the type should probably be label

This is the section I commented on:

```
let rec: MultinetType = 'label';
if (isKey) {
  rec = 'label';
}
```

I'm saying the `if` statement there is redundant, as `'label'` is already the default value of `rec`.

Reply:
`rec` was meant to be short for "recommendation", but `recommendation` seems too long for a variable name. If it's needed I can use `recommendation` or a different name for this variable.

I kept the `isKey` test for legibility. In pre-typescript days, I wouldn't have assigned any value to `rec` in its initializer, and then the if-elseif-else chain would be explicit and non-redundant. My choices with typescript are to (1) leave `rec` undefined and then type it as `MultinetType | undefined`; (2) cast the if-elseif-else chain as a function that gets called in-place (replacing the `rec =` parts with `return`) and assigned to `rec`; or (3) initialize the variable because I have to, and then use the expclicit if-elseif-else chain as I did.

(1) is a bad choice because it's going out of its way to do something simple.
(2) is a bad choice because it's confusing/ugly.
(3) is non-optimal because it looks redundant. HOWEVER, I think "optimizing" this code would be worse: it would leave it unclear that we deliberately select `label` for Multinet key fields.

This is another case of explicit code for the goal of comprehensibility. Let me know if this is ok or if you really want a different solution here.

Reply:
I was missing the fact that the first `if` statement (that I claimed was redundant) has an effect, which is that it sets the priority of the checked conditions (`isKey` is the first checked condition, and if true will ensure that no other conditional statement overwrites `rec`). You could change the redundant assignment with the code below, but I'm also okay with keeping it as is. Up to you.
```
            let rec: MultinetType = 'label';
            if (!isKey) {
              if (category && !number && !date) {
                rec = 'category';
              } else if (number) {
                rec = 'number';
              } else if (date) {
                rec = 'date';
              }
            }
```

Reply:
I sort of like this approach but in the interest of keeping things moving, let's leave the code as is.